lib/floe/workflow/runner/kubernetes.rb

```diff
@@ -20,26 +20,49 @@ def initialize(options = {})
           super
         end
 
-        def run!(resource, env = {}, secrets = {})
+        def run_async!(resource, env = {}, secrets = {})
           raise ArgumentError, "Invalid resource" unless resource&.start_with?("docker://")
 
           image     = resource.sub("docker://", "")
-          name      = pod_name(image)
+          pod       = pod_name(image)
           secret    = create_secret!(secrets) unless secrets&.empty?
           overrides = pod_spec(image, env, secret)
 
-          result = kubectl_run!(image, name, overrides)
+          begin
+            kubectl_run!(image, pod, overrides)
+          rescue
+            delete_secret(secret) if secret
+            raise
+          end
+
+          [pod, secret]
+        end
+
+        def running?(pod_name)
+          %w[Pending Running].include?(pod_info(pod_name).dig("status", "phase"))
+        end
+
+        def success?(pod_name)
+          pod_info(pod_name).dig("status", "phase") == "Succeeded"
+        end
 
-          # Kubectl prints that the pod was deleted, strip this from the output
-          output = result.output.gsub(/pod "#{name}" deleted/, "")
+        def output(pod)
+          result = kubectl!("logs", pod, [:namespace, namespace])
+          result.output
+        end
 
-          [result.exit_status, output]
-        ensure
-          delete_secret!(secret) if secret
+        def cleanup(pod, secret)
+          delete_pod(pod)       if pod
+          delete_secret(secret) if secret
         end
 
         private
 
+        def pod_info(pod_name)
+          results = kubectl!("get", "pod", pod_name, [:output, :json], [:namespace, namespace])
```

Comment:
Would a watch be better or is that overkill?

Reply:
Oh we're calling kubectl directly - thought we were using something like kubeclient. 🤔

Reply:
So I'm currently using `run` to call the async methods in order to confirm that async behaves the same as sync but it would be more efficient to leave `run` as a blocking call using fewer API calls.

For anything orchestrating >1 workflow run then a watch on all pods would absolutely be the way to go.